File: Algorithms/0051_N-Queens/Python/N-Queens_Solution_1.py

# Space: O(N)
# Time: O(n)

class Solution:
    def solveNQueens(self, n):
        self.queen_position = []
        self.res = []

        def backtracking(row_num, queen_count, required_queen, length):
            if queen_count == required_queen:
                self.res.append(self.queen_position[:])
                return

            for x in range(length):
                if self.check_position(x, row_num, length, self.queen_position):
                    self.queen_position.append([row_num, x])
                    backtracking(row_num + 1, queen_count + 1, required_queen, length)
                    self.queen_position.pop()

        backtracking(0, 0, n, n)

        return self.generate_res(self.res, n)

    def check_position(self, x, y, length, alist):
        if len(alist) == 0: return True

        for queen_y, queen_x in alist:
            if x == queen_x or y == queen_y: return False

        direction = [[1, 1], [-1, -1], [-1, 1], [1, -1]]

        for dir_y, dir_x in direction:
            temp_x, temp_y = x, y

            while 0 <= temp_x < length and 0 <= temp_y < length:
                if [temp_y, temp_x] in alist: return False
                temp_x += dir_x
                temp_y += dir_y

        return True

    def generate_res(self, alist, length):
        res = []
        for each_res in alist:
            temp = ['.' * length] * length
            for y, x in each_res:
                temp[y] = temp[y][:x] + 'Q' + temp[y][x + 1:]
            res.append(temp)
        return res

File: Algorithms/0051_N-Queens/Java/N-QueensSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    List<List<String>> res = new ArrayList<>();
    int length = 0;

    public List<List<String>> solveNQueens(int n) {
        if (n == 1) {
            List<List<String>> res = new ArrayList<>();
            res.add(new ArrayList<>(Arrays.asList("Q")));
            return res;
        }
        length = n;
        String[][] board = new String[n][n];
        for (int row = 0; row < board.length; row++) {
            Arrays.fill(board[row], ".");
        }
        dfs(board, n, 0);
        return res;
    }

    private List<String> generateRes(String[][] board) {
        List<String> res = new ArrayList<>();
        for (int y = 0; y < board.length; y++) {
            res.add(Arrays.stream(board[y]).map(Objects::toString).collect(Collectors.joining()));
        }
        return res;
    }

    private void dfs(String[][] board, int queens, int y) {
        if (queens == 0) {
            List<String> tempRes = generateRes(board);
            res.add(tempRes);
            return;
        }
        for (int i = 0; i < length; i++) {
            if (verify(board, y, i)) {
                queens--;
                board[y][i] = "Q";
                dfs(board, queens, y + 1);
                board[y][i] = ".";
                queens++;
            }
        }
    }

    private boolean verify(String[][] board, int y, int x) {
        int tempY = 0, tempX = 0;
        // check top
        tempY = y;
        while (tempY >= 0) {
            if (board[tempY][x] == "Q") return false;
            tempY--;
        }
        // check diagonal
        tempX = x;
        tempY = y;
        while (tempY >= 0 && tempX >= 0) {
            if (board[tempY][tempX] == "Q") return false;
            tempY--;
            tempX--;
        }

        tempX = x;
        tempY = y;
        while (tempY >= 0 && tempX < length) {
            if (board[tempY][tempX] == "Q") return false;
            tempY--;
            tempX++;
        }
        return true;
    }
}




